relays/ethereum/src/rpc.rs

```diff
@@ -0,0 +1,338 @@
+	async fn block_hash_by_number(&mut self, number: SubBlockNumber) -> Result<SubstrateHash>;
+	async fn header_by_number(&mut self, block_number: SubBlockNumber) -> Result<SubstrateHeader>;
+	async fn next_account_index(&mut self, account: node_primitives::AccountId) -> Result<node_primitives::Index>;
+	async fn best_ethereum_block(&mut self) -> Result<EthereumHeaderId>;
+	async fn ethereum_receipts_required(&mut self, header: SubstrateEthereumHeader) -> Result<bool>;
+	async fn ethereum_header_known(&mut self, header_id: EthereumHeaderId) -> Result<bool>;
+	async fn submit_extrinsic(&mut self, transaction: UncheckedExtrinsic) -> Result<SubstrateHash>;
+	async fn grandpa_authorities_set(&mut self, block: SubstrateHash) -> Result<GrandpaAuthorityList>;
+}
+
+/// The client used to interact with a Substrate node through RPC.
+pub struct SubstrateRpcClient {
+	client: RawClient<HttpTransportClient>,
+}
+
+impl SubstrateRpcClient {
+	/// Create a new Substrate RPC Client.
+	pub fn new(params: SubstrateConnectionParams) -> Self {
+		let uri = format!("http://{}:{}", params.host, params.port);
+		let transport = HttpTransportClient::new(&uri);
+		let client = RawClient::new(transport);
+
+		Self { client }
+	}
+}
+
+#[async_trait]
+impl SubstrateRpc for SubstrateRpcClient {
+	/// Returns the best Substrate header.
+	async fn best_header(&mut self) -> Result<SubstrateHeader> {
+		Ok(Substrate::chain_get_header(&mut self.client, Params::None).await?)
+	}
+
+	/// Get a Substrate header by its hash.
+	async fn header_by_hash(&mut self, hash: SubstrateHash) -> Result<SubstrateHeader> {
+		let hash = serde_json::to_value(hash)?;
+		let params = Params::Array(vec![hash]);
+
+		Ok(Substrate::chain_get_header(&mut self.client, params).await?)
+	}
+
+	/// Get a Substrate block hash by its number.
+	async fn block_hash_by_number(&mut self, number: SubBlockNumber) -> Result<SubstrateHash> {
+		let params = Params::Array(vec![serde_json::to_value(number)?]);
+
+		Ok(Substrate::chain_get_block_hash(&mut self.client, params).await?)
+	}
+
+	/// Get a Substrate header by its number.
+	async fn header_by_number(&mut self, block_number: SubBlockNumber) -> Result<SubstrateHeader> {
+		let block_hash = Self::block_hash_by_number(self, block_number).await?;
+		Ok(Self::header_by_hash(self, block_hash).await?)
+	}
+
+	/// Get the nonce of the given Substrate account.
+	async fn next_account_index(&mut self, account: node_primitives::AccountId) -> Result<node_primitives::Index> {
+		// Q: Should this belong here, or be left to the caller?
+		use sp_core::crypto::Ss58Codec;
+		let account = serde_json::to_value(account.to_ss58check())?;
+		let params = Params::Array(vec![account]);
+
+		Ok(Substrate::system_account_next_index(&mut self.client, params).await?)
+	}
+
+	/// Returns best Ethereum block that Substrate runtime knows of.
+	async fn best_ethereum_block(&mut self) -> Result<EthereumHeaderId> {
+		let call = EthereumHeadersApiCalls::BestBlock.to_string();
```

Comment:
I feel like this is a bit higher level call and my initial idea was to only extract low-level methods (like `state_call`) into a separate module and keep the ones that are more bridge-specific in a separate place. I don't feel strong about it though, but we should pay attention to avoid putting to much bridge-specific logic in this module.

Reply:
I thought it would be nice for the module to encapsulate the parameter encoding/decoding aspect of it so the caller wouldn't have to worry about that themselves. These mostly what I consider "base" calls in that they're a wrapper around a single RPC method. There will be other high level calls, like ones which interact with contracts, which will be elsewhere (like in `ethereum_client.rs`). I'm open to moving these elsewhere, but I think that can be done later unless you have a strong opinion on why it should be done now.